### ISSUE TITLE
Which internal properties should we store in sparse format?

### ISSUE THREAD

**Issue by jdbancal** [replab/replab#285]
The property `self.V` of `replab.bsgs.ChainWithImages` seems to be stored in a non-sparse manner. As a result, the call to `self.J.compose` in https://github.com/replab/replab/blob/fd3d8d992dd151c68d752e721b1daabd8ee156df/src/%2Breplab/%2Bbsgs/ChainWithImages.m#L464 is un-necessarily slow.

For instance, the decomposition of the following 937-dimensional representation of the fully symmetric group S(6) takes 8 sec, 4 of which are spent doing these multiplications (which would take virtually no time in sparse format).

```
gens{1} = [1  782  783  808  809  810  811  812  813  784  789  785  786  787  788  790  795  791  792  793  794  796  801  797  798  799  800  802  807  803  804  805  806  814  839  840  841  842  843  844  815  820  816  817  818  819  821  826  822  823  824  825  827  832  828  829  830  831  833  838  834  835  836  837  845  870  871  872  873  874  875  846  851  847  848  849  850  852  857  853  854  855  856  858  863  859  860  861  862  864  869  865  866  867  868  876  901  902  903  904  905  906  877  882  878  879  880  881  883  888  884  885  886  887  889  894  890  891  892  893  895  900  896  897  898  899  907  932  933  934  935  936  937  908  913  909  910  911  912  914  919  915  916  917  918  920  925  921  922  923  924  926  931  927  928  929  930    2  127  128  133  129  130  131  132  134  139  135  136  137  138  140  145  141  142  143  144  146  151  147  148  149  150  152  157  153  154  155  156    3   28   29   30   31   32   33    4    9    5    6    7    8   10   15   11   12   13   14   16   21   17   18   19   20   22   27   23   24   25   26   34   59   60   61   62   63   64   35   40   36   37   38   39   41   46   42   43   44   45   47   52   48   49   50   51   53   58   54   55   56   57   65   90   91   92   93   94   95   66   71   67   68   69   70   72   77   73   74   75   76   78   83   79   80   81   82   84   89   85   86   87   88   96  121  122  123  124  125  126   97  102   98   99  100  101  103  108  104  105  106  107  109  114  110  111  112  113  115  120  116  117  118  119  158  283  284  289  285  286  287  288  290  295  291  292  293  294  296  301  297  298  299  300  302  307  303  304  305  306  308  313  309  310  311  312  159  184  185  186  187  188  189  160  165  161  162  163  164  166  171  167  168  169  170  172  177  173  174  175  176  178  183  179  180  181  182  190  215  216  217  218  219  220  191  196  192  193  194  195  197  202  198  199  200  201  203  208  204  205  206  207  209  214  210  211  212  213  221  246  247  248  249  250  251  222  227  223  224  225  226  228  233  229  230  231  232  234  239  235  236  237  238  240  245  241  242  243  244  252  277  278  279  280  281  282  253  258  254  255  256  257  259  264  260  261  262  263  265  270  266  267  268  269  271  276  272  273  274  275  314  439  440  445  441  442  443  444  446  451  447  448  449  450  452  457  453  454  455  456  458  463  459  460  461  462  464  469  465  466  467  468  315  340  341  342  343  344  345  316  321  317  318  319  320  322  327  323  324  325  326  328  333  329  330  331  332  334  339  335  336  337  338  346  371  372  373  374  375  376  347  352  348  349  350  351  353  358  354  355  356  357  359  364  360  361  362  363  365  370  366  367  368  369  377  402  403  404  405  406  407  378  383  379  380  381  382  384  389  385  386  387  388  390  395  391  392  393  394  396  401  397  398  399  400  408  433  434  435  436  437  438  409  414  410  411  412  413  415  420  416  417  418  419  421  426  422  423  424  425  427  432  428  429  430  431  470  595  596  601  597  598  599  600  602  607  603  604  605  606  608  613  609  610  611  612  614  619  615  616  617  618  620  625  621  622  623  624  471  496  497  498  499  500  501  472  477  473  474  475  476  478  483  479  480  481  482  484  489  485  486  487  488  490  495  491  492  493  494  502  527  528  529  530  531  532  503  508  504  505  506  507  509  514  510  511  512  513  515  520  516  517  518  519  521  526  522  523  524  525  533  558  559  560  561  562  563  534  539  535  536  537  538  540  545  541  542  543  544  546  551  547  548  549  550  552  557  553  554  555  556  564  589  590  591  592  593  594  565  570  566  567  568  569  571  576  572  573  574  575  577  582  578  579  580  581  583  588  584  585  586  587  626  751  752  757  753  754  755  756  758  763  759  760  761  762  764  769  765  766  767  768  770  775  771  772  773  774  776  781  777  778  779  780  627  652  653  654  655  656  657  628  633  629  630  631  632  634  639  635  636  637  638  640  645  641  642  643  644  646  651  647  648  649  650  658  683  684  685  686  687  688  659  664  660  661  662  663  665  670  666  667  668  669  671  676  672  673  674  675  677  682  678  679  680  681  689  714  715  716  717  718  719  690  695  691  692  693  694  696  701  697  698  699  700  702  707  703  704  705  706  708  713  709  710  711  712  720  745  746  747  748  749  750  721  726  722  723  724  725  727  732  728  729  730  731  733  738  734  735  736  737  739  744  740  741  742  743];
gens{2} = [1  158  159  160  161  162  163  164  165  166  168  167  169  170  171  172  174  173  175  176  177  178  180  179  181  182  183  184  186  185  187  188  189  190  197  198  199  200  201  202  191  192  193  194  195  196  203  205  204  206  207  208  209  211  210  212  213  214  215  217  216  218  219  220  221  228  229  230  231  232  233  222  223  224  225  226  227  234  236  235  237  238  239  240  242  241  243  244  245  246  248  247  249  250  251  252  259  260  261  262  263  264  253  254  255  256  257  258  265  267  266  268  269  270  271  273  272  274  275  276  277  279  278  280  281  282  283  290  291  292  293  294  295  284  285  286  287  288  289  296  298  297  299  300  301  302  304  303  305  306  307  308  310  309  311  312  313    2    3    4    5    6    7    8    9   10   12   11   13   14   15   16   18   17   19   20   21   22   24   23   25   26   27   28   30   29   31   32   33   34   41   42   43   44   45   46   35   36   37   38   39   40   47   49   48   50   51   52   53   55   54   56   57   58   59   61   60   62   63   64   65   72   73   74   75   76   77   66   67   68   69   70   71   78   80   79   81   82   83   84   86   85   87   88   89   90   92   91   93   94   95   96  103  104  105  106  107  108   97   98   99  100  101  102  109  111  110  112  113  114  115  117  116  118  119  120  121  123  122  124  125  126  127  134  135  136  137  138  139  128  129  130  131  132  133  140  142  141  143  144  145  146  148  147  149  150  151  152  154  153  155  156  157  314  346  347  348  349  350  351  352  353  355  354  356  357  358  359  361  360  362  363  364  365  367  366  368  369  370  371  373  372  374  375  376  315  316  317  318  319  320  321  322  324  323  325  326  327  328  330  329  331  332  333  334  336  335  337  338  339  340  342  341  343  344  345  377  384  385  386  387  388  389  378  379  380  381  382  383  390  392  391  393  394  395  396  398  397  399  400  401  402  404  403  405  406  407  408  415  416  417  418  419  420  409  410  411  412  413  414  421  423  422  424  425  426  427  429  428  430  431  432  433  435  434  436  437  438  439  446  447  448  449  450  451  440  441  442  443  444  445  452  454  453  455  456  457  458  460  459  461  462  463  464  466  465  467  468  469  470  502  503  504  505  506  507  508  509  511  510  512  513  514  515  517  516  518  519  520  521  523  522  524  525  526  527  529  528  530  531  532  471  472  473  474  475  476  477  478  480  479  481  482  483  484  486  485  487  488  489  490  492  491  493  494  495  496  498  497  499  500  501  533  540  541  542  543  544  545  534  535  536  537  538  539  546  548  547  549  550  551  552  554  553  555  556  557  558  560  559  561  562  563  564  571  572  573  574  575  576  565  566  567  568  569  570  577  579  578  580  581  582  583  585  584  586  587  588  589  591  590  592  593  594  595  602  603  604  605  606  607  596  597  598  599  600  601  608  610  609  611  612  613  614  616  615  617  618  619  620  622  621  623  624  625  626  658  659  660  661  662  663  664  665  667  666  668  669  670  671  673  672  674  675  676  677  679  678  680  681  682  683  685  684  686  687  688  627  628  629  630  631  632  633  634  636  635  637  638  639  640  642  641  643  644  645  646  648  647  649  650  651  652  654  653  655  656  657  689  696  697  698  699  700  701  690  691  692  693  694  695  702  704  703  705  706  707  708  710  709  711  712  713  714  716  715  717  718  719  720  727  728  729  730  731  732  721  722  723  724  725  726  733  735  734  736  737  738  739  741  740  742  743  744  745  747  746  748  749  750  751  758  759  760  761  762  763  752  753  754  755  756  757  764  766  765  767  768  769  770  772  771  773  774  775  776  778  777  779  780  781  782  814  815  816  817  818  819  820  821  823  822  824  825  826  827  829  828  830  831  832  833  835  834  836  837  838  839  841  840  842  843  844  783  784  785  786  787  788  789  790  792  791  793  794  795  796  798  797  799  800  801  802  804  803  805  806  807  808  810  809  811  812  813  845  852  853  854  855  856  857  846  847  848  849  850  851  858  860  859  861  862  863  864  866  865  867  868  869  870  872  871  873  874  875  876  883  884  885  886  887  888  877  878  879  880  881  882  889  891  890  892  893  894  895  897  896  898  899  900  901  903  902  904  905  906  907  914  915  916  917  918  919  908  909  910  911  912  913  920  922  921  923  924  925  926  928  927  929  930  931  932  934  933  935  936  937];
G = replab.S(937).subgroup(gens);
rep = G.naturalRep;
tic;
dec = rep.decomposition;
toc;
```

Hence two questions:
 1. Are all sparse internal object stored in a sparse manner?
 2. In the case of the present example, would there be a more clever way to obtain the decomposition of this representation? Maybe by providing a BSGS chain for S(6), which would be very short (hence dense encoding would have a negligible effect)?

**Comment by jdbancal** [comment]
If I try to construct the representation by hand from S(6) and with sparse images only, internal computations still involve dense matrices
```
% Constructing the small group
gensS6{1} = [2:6 1];
gensS6{2} = [2 1 3:6];
G6 = replab.S(6).subgroup(gensS6); % to make sure the generator ordering is as desired

% Defining the images
gens{1} = [1  782  783  808  809  810  811  812  813  784  789  785  786  787  788  790  795  791  792  793  794  796  801  797  798  799  800  802  807  803  804  805  806  814  839  840  841  842  843  844  815  820  816  817  818  819  821  826  822  823  824  825  827  832  828  829  830  831  833  838  834  835  836  837  845  870  871  872  873  874  875  846  851  847  848  849  850  852  857  853  854  855  856  858  863  859  860  861  862  864  869  865  866  867  868  876  901  902  903  904  905  906  877  882  878  879  880  881  883  888  884  885  886  887  889  894  890  891  892  893  895  900  896  897  898  899  907  932  933  934  935  936  937  908  913  909  910  911  912  914  919  915  916  917  918  920  925  921  922  923  924  926  931  927  928  929  930    2  127  128  133  129  130  131  132  134  139  135  136  137  138  140  145  141  142  143  144  146  151  147  148  149  150  152  157  153  154  155  156    3   28   29   30   31   32   33    4    9    5    6    7    8   10   15   11   12   13   14   16   21   17   18   19   20   22   27   23   24   25   26   34   59   60   61   62   63   64   35   40   36   37   38   39   41   46   42   43   44   45   47   52   48   49   50   51   53   58   54   55   56   57   65   90   91   92   93   94   95   66   71   67   68   69   70   72   77   73   74   75   76   78   83   79   80   81   82   84   89   85   86   87   88   96  121  122  123  124  125  126   97  102   98   99  100  101  103  108  104  105  106  107  109  114  110  111  112  113  115  120  116  117  118  119  158  283  284  289  285  286  287  288  290  295  291  292  293  294  296  301  297  298  299  300  302  307  303  304  305  306  308  313  309  310  311  312  159  184  185  186  187  188  189  160  165  161  162  163  164  166  171  167  168  169  170  172  177  173  174  175  176  178  183  179  180  181  182  190  215  216  217  218  219  220  191  196  192  193  194  195  197  202  198  199  200  201  203  208  204  205  206  207  209  214  210  211  212  213  221  246  247  248  249  250  251  222  227  223  224  225  226  228  233  229  230  231  232  234  239  235  236  237  238  240  245  241  242  243  244  252  277  278  279  280  281  282  253  258  254  255  256  257  259  264  260  261  262  263  265  270  266  267  268  269  271  276  272  273  274  275  314  439  440  445  441  442  443  444  446  451  447  448  449  450  452  457  453  454  455  456  458  463  459  460  461  462  464  469  465  466  467  468  315  340  341  342  343  344  345  316  321  317  318  319  320  322  327  323  324  325  326  328  333  329  330  331  332  334  339  335  336  337  338  346  371  372  373  374  375  376  347  352  348  349  350  351  353  358  354  355  356  357  359  364  360  361  362  363  365  370  366  367  368  369  377  402  403  404  405  406  407  378  383  379  380  381  382  384  389  385  386  387  388  390  395  391  392  393  394  396  401  397  398  399  400  408  433  434  435  436  437  438  409  414  410  411  412  413  415  420  416  417  418  419  421  426  422  423  424  425  427  432  428  429  430  431  470  595  596  601  597  598  599  600  602  607  603  604  605  606  608  613  609  610  611  612  614  619  615  616  617  618  620  625  621  622  623  624  471  496  497  498  499  500  501  472  477  473  474  475  476  478  483  479  480  481  482  484  489  485  486  487  488  490  495  491  492  493  494  502  527  528  529  530  531  532  503  508  504  505  506  507  509  514  510  511  512  513  515  520  516  517  518  519  521  526  522  523  524  525  533  558  559  560  561  562  563  534  539  535  536  537  538  540  545  541  542  543  544  546  551  547  548  549  550  552  557  553  554  555  556  564  589  590  591  592  593  594  565  570  566  567  568  569  571  576  572  573  574  575  577  582  578  579  580  581  583  588  584  585  586  587  626  751  752  757  753  754  755  756  758  763  759  760  761  762  764  769  765  766  767  768  770  775  771  772  773  774  776  781  777  778  779  780  627  652  653  654  655  656  657  628  633  629  630  631  632  634  639  635  636  637  638  640  645  641  642  643  644  646  651  647  648  649  650  658  683  684  685  686  687  688  659  664  660  661  662  663  665  670  666  667  668  669  671  676  672  673  674  675  677  682  678  679  680  681  689  714  715  716  717  718  719  690  695  691  692  693  694  696  701  697  698  699  700  702  707  703  704  705  706  708  713  709  710  711  712  720  745  746  747  748  749  750  721  726  722  723  724  725  727  732  728  729  730  731  733  738  734  735  736  737  739  744  740  741  742  743];
gens{2} = [1  158  159  160  161  162  163  164  165  166  168  167  169  170  171  172  174  173  175  176  177  178  180  179  181  182  183  184  186  185  187  188  189  190  197  198  199  200  201  202  191  192  193  194  195  196  203  205  204  206  207  208  209  211  210  212  213  214  215  217  216  218  219  220  221  228  229  230  231  232  233  222  223  224  225  226  227  234  236  235  237  238  239  240  242  241  243  244  245  246  248  247  249  250  251  252  259  260  261  262  263  264  253  254  255  256  257  258  265  267  266  268  269  270  271  273  272  274  275  276  277  279  278  280  281  282  283  290  291  292  293  294  295  284  285  286  287  288  289  296  298  297  299  300  301  302  304  303  305  306  307  308  310  309  311  312  313    2    3    4    5    6    7    8    9   10   12   11   13   14   15   16   18   17   19   20   21   22   24   23   25   26   27   28   30   29   31   32   33   34   41   42   43   44   45   46   35   36   37   38   39   40   47   49   48   50   51   52   53   55   54   56   57   58   59   61   60   62   63   64   65   72   73   74   75   76   77   66   67   68   69   70   71   78   80   79   81   82   83   84   86   85   87   88   89   90   92   91   93   94   95   96  103  104  105  106  107  108   97   98   99  100  101  102  109  111  110  112  113  114  115  117  116  118  119  120  121  123  122  124  125  126  127  134  135  136  137  138  139  128  129  130  131  132  133  140  142  141  143  144  145  146  148  147  149  150  151  152  154  153  155  156  157  314  346  347  348  349  350  351  352  353  355  354  356  357  358  359  361  360  362  363  364  365  367  366  368  369  370  371  373  372  374  375  376  315  316  317  318  319  320  321  322  324  323  325  326  327  328  330  329  331  332  333  334  336  335  337  338  339  340  342  341  343  344  345  377  384  385  386  387  388  389  378  379  380  381  382  383  390  392  391  393  394  395  396  398  397  399  400  401  402  404  403  405  406  407  408  415  416  417  418  419  420  409  410  411  412  413  414  421  423  422  424  425  426  427  429  428  430  431  432  433  435  434  436  437  438  439  446  447  448  449  450  451  440  441  442  443  444  445  452  454  453  455  456  457  458  460  459  461  462  463  464  466  465  467  468  469  470  502  503  504  505  506  507  508  509  511  510  512  513  514  515  517  516  518  519  520  521  523  522  524  525  526  527  529  528  530  531  532  471  472  473  474  475  476  477  478  480  479  481  482  483  484  486  485  487  488  489  490  492  491  493  494  495  496  498  497  499  500  501  533  540  541  542  543  544  545  534  535  536  537  538  539  546  548  547  549  550  551  552  554  553  555  556  557  558  560  559  561  562  563  564  571  572  573  574  575  576  565  566  567  568  569  570  577  579  578  580  581  582  583  585  584  586  587  588  589  591  590  592  593  594  595  602  603  604  605  606  607  596  597  598  599  600  601  608  610  609  611  612  613  614  616  615  617  618  619  620  622  621  623  624  625  626  658  659  660  661  662  663  664  665  667  666  668  669  670  671  673  672  674  675  676  677  679  678  680  681  682  683  685  684  686  687  688  627  628  629  630  631  632  633  634  636  635  637  638  639  640  642  641  643  644  645  646  648  647  649  650  651  652  654  653  655  656  657  689  696  697  698  699  700  701  690  691  692  693  694  695  702  704  703  705  706  707  708  710  709  711  712  713  714  716  715  717  718  719  720  727  728  729  730  731  732  721  722  723  724  725  726  733  735  734  736  737  738  739  741  740  742  743  744  745  747  746  748  749  750  751  758  759  760  761  762  763  752  753  754  755  756  757  764  766  765  767  768  769  770  772  771  773  774  775  776  778  777  779  780  781  782  814  815  816  817  818  819  820  821  823  822  824  825  826  827  829  828  830  831  832  833  835  834  836  837  838  839  841  840  842  843  844  783  784  785  786  787  788  789  790  792  791  793  794  795  796  798  797  799  800  801  802  804  803  805  806  807  808  810  809  811  812  813  845  852  853  854  855  856  857  846  847  848  849  850  851  858  860  859  861  862  863  864  866  865  867  868  869  870  872  871  873  874  875  876  883  884  885  886  887  888  877  878  879  880  881  882  889  891  890  892  893  894  895  897  896  898  899  900  901  903  902  904  905  906  907  914  915  916  917  918  919  908  909  910  911  912  913  920  922  921  923  924  925  926  928  927  929  930  931  932  934  933  935  936  937];
image1 = sparse(gens{1}, [1:937], 1);
image2 = sparse(gens{2}, [1:937], 1);
image1inv = sparse([1:937], gens{1}, 1);
image2inv = sparse([1:937], gens{2}, 1);

% Construct the representation
rep = G6.repByImages('R', 937, {image1, image2}, {image1inv image2inv});

% Decompose the representation
rep.decomposition
```

**Comment by denisrosset** [comment]
We should preserve sparsity. What happens here is that the identity matrix is dense, and then that format propagates.

This is being fixed by #309
